Machine Learning Zoomcamp
=========================


Email: ad4c1c7fac2949764e8b27b012982dfcd7cfa9eb



In [1]:
import pandas as pd
import numpy as np

### Question 1  
What's the version of NumPy that you installed?

You can get the version information using the __version__ field:

In [2]:
np.__version__

'1.23.1'

### A1: 
1.23.1

### Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

In [3]:
df_cars = (
    pd.read_csv('data/car_data_set_zc/data.csv')
)
df_cars.shape[0]
    

11914

### A2:
11914

### Question 3
Who are the most popular car manufacturers (top-3) according to the dataset?

In [4]:
print(", ".join(
    df_cars.groupby(['Make']).size()
    .sort_values(ascending=False)
    .head(3)
    .reset_index()
    ['Make']
    .to_list()
))

Chevrolet, Ford, Volkswagen


A3:  
Chevrolet, Ford, Volkswagen

### Question 4
What's the number of unique Audi car models in the dataset?

In [5]:
(
    df_cars[df_cars['Make'] == 'Audi'].groupby(['Model']).size().count()
)

34

### A4
34


In [6]:
### Question 5
How many columns in the dataset have missing values?


Object `values` not found.


In [7]:
(
    df_cars.isnull().sum()
    .reset_index()
    .rename(columns={0: 'missing_values'})
    .query('missing_values > 0')
)

index  missing_values
3  Engine Fuel Type               3
4         Engine HP              69
5  Engine Cylinders              30
8   Number of Doors               6
9   Market Category            3742

### A5
5

### Question 6
Find the median value of "Engine Cylinders" column in the dataset.  
Next, calculate the most frequent value of the same "Engine Cylinders".  
Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.  
Now, calculate the median value of "Engine Cylinders" once again.  

In [8]:
ec_median = df_cars['Engine Cylinders'].median()
ec_mode = df_cars['Engine Cylinders'].mode()[0]
ec_mean = df_cars['Engine Cylinders'].mean()

df_cars = df_cars.fillna({'Engine Cylinders': ec_mode})

ec_median2 = df_cars['Engine Cylinders'].median()
ec_mean2 = df_cars['Engine Cylinders'].mean()



print(f"{ec_median=}")
print(f"{ec_mode=}")
print(f"{ec_mean=}")

print(f"{ec_median2=}")
print(f"{ec_mean2=}")

ec_median=6.0
ec_mode=4.0
ec_mean=5.628828677213059
ec_median2=6.0
ec_mean2=5.624727211683734


### A6)
No

### Question 7
Select all the "Lotus" cars from the dataset.  
Select only columns "Engine HP", "Engine Cylinders".  
Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).  
Get the underlying NumPy array. Let's call it X.  
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.  
Invert XTX.  
Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].  
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.  
What's the value of the first element of w?  


In [35]:
X = (
    df_cars
    [df_cars['Make'] == 'Lotus']
    [['Engine HP', 'Engine Cylinders']]
    .drop_duplicates()
    .to_numpy()
)

XTX = X.T.dot(X)
XTX_inv = np.linalg.inv(XTX)

y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

w = XTX_inv.dot(X.T).dot(y)
print(w)

[  4.59494481 -63.56432501]


In [ ]:
### A7
4.59494481